File: algorithms/113. Path Sum II/solution.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> result;
        vector<int> path;
        dfs(root, sum, path, result);
        return result;
    }

    void dfs(TreeNode * root, int target, vector<int> & path, vector<vector<int>> & result) {
        if ( ! root ) return;
        path.push_back(root->val);
        if ( ! root->left && ! root->right && root->val == target) result.push_back(path);
        dfs(root->left, target - root->val, path, result);
        dfs(root->right, target - root->val, path, result);
        path.pop_back();
    }
};

File: algorithms/113. Path Sum II/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        def dfs(root, target):
            if not root:
                return
            if not root.left and not root.right:
                if target == root.val:
                    res.append(path + [root.val])
                return
            path.append(root.val)
            dfs(root.left, target - root.val)
            dfs(root.right, target - root.val)
            path.pop()
        res, path = [], []
        dfs(root, sum)
        return res

# 114 / 114 test cases passed.
# Status: Accepted
# Runtime: 66 ms
# beats 96.94 %
